### ISSUE TITLE
Consider making Minkowski distance a parametric type

### ISSUE THREAD

**Issue by ahwillia** [JuliaStats/Distances.jl#56]
It would be nice to be able to dispatch on the `p` parameter in the `Minkowski` distance type. There is a very similar implementation here (thanks to @Evizero), which I think could be ported here if others are interested in this functionality: https://github.com/JuliaML/Losses.jl/blob/master/src/supervised/distance.jl#L10

**Comment by ararslan** [comment]
Currently it's defined as
```julia
immutable Minkowski{T <: Real} <: Metric
    p::T
end
```
So I think what you're after is already implemented?

**Comment by ahwillia** [comment]
No I'm after something slightly different. 

```julia
using LossFunctions

x = LPDistLoss(2) # same as L2DistLoss
y = LPDistLoss(1) # same as L1DistLoss

# multiple dispatch works for different norms!
foo(::LPDistLoss{1}) = ...
foo(::LPDistLoss{2}) = ...
```

The idea being that two `Minkowski` types with different values for `p`'s are really different distance metrics and each should have its own type.

**Comment by KristofferC** [comment]
Yes, having the exponent as a type parameter seems like a good idea.

**Comment by dkarrasch** [comment]
I think that along the same lines, one could have a single interface for all `UnionMetrics`, including the weighted/periodic ones? By including the extra info as a type parameter, we could pass the current index `i` to the `eval_op`, and inside there get the dimension parameter (weight/period) with the help of the index `i`. Does that sound reasonable? See #107.
